crypto/ec/ec_kmeth.c

```diff
@@ -168,6 +168,8 @@ EC_KEY *EC_KEY_new_method(ENGINE *engine)
     ret->references = 1;
 
     if (ret->meth->init != NULL && ret->meth->init(ret) == 0) {
+        CRYPTO_free_ex_data(CRYPTO_EX_INDEX_EC_KEY, ret, &ret->ex_data);
+        CRYPTO_THREAD_lock_free(ret->lock);
```

Comment:
Do we need to call `ENGINE_finish()` (guarded by OPENSSL_NO_ENGINE) like the other failure case above? It seems to me there is a lot of duplicated clean up code in this function which could be pulled into a separate `goto` block at the end.